### PR TITLE
feat: match no_sync with pytorch

### PR DESCRIPTION
**Pull request by k223kim**
<details>
  <summary><b>Before submitting</b></summary>

- [ ] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [ ] Did you read the [contributor guideline](https://github.com/Lightning-AI/pytorch-lightning/blob/main/.github/CONTRIBUTING.md), Pull Request section?
- [ ] Did you make sure to update the docs?
- [ ] Did you write any new necessary tests?

</details>

## What does this PR do?

Fixes # (issue).

## PR review

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: thunder/core/module.py [hunk 1]
```diff
@@ -379,8 +379,8 @@ def no_sync(self):
             yield
         finally:
             reset_skip_data_parallel_grad_sync(token)
-            _sync_grads(self)
 
+    #
```
Comment (Borda) [inline thunder/core/module.py]:
```suggestion
```

File: thunder/tests/distributed/helper.py [hunk 2]
```diff
@@ -431,3 +431,109 @@ def get_ground_truth_loss_grads(device, dataloader):
...
+                if num_grad_accum_steps is not None:
+                    loss /= num_grad_accum_steps
+                loss.backward()
+
+            return loss
+
+        device = torch.device("cuda", test_case.rank)
+
+        gradients = defaultdict(list)
+        for use_no_sync in (True, False):
```
Comment (Borda) [inline thunder/tests/distributed/helper.py]: can we have it rather as an argument ot we really need this loop?

Comment (k223kim) [reply thunder/tests/distributed/helper.py]: Hmm, we do not need this loop. Would it make a difference if we use arguments?

File: thunder/tests/distributed/helper.py [hunk 3]
```diff
@@ -431,3 +431,109 @@ def get_ground_truth_loss_grads(device, dataloader):
...
+                            )
+                        if use_no_sync and i == 0 and iter_count == 0:
+                            import thunder
+
+                            # make sure the backward trace under `no_sync` has actual math computations.
+                            no_sync_bwd_trc = thunder.last_backward_traces(jitted_model)[-1]
+                            test_case.assertGreater(len(no_sync_bwd_trc.bound_symbols), 1)
+                assert torch.allclose(torch_loss, loss, atol=1e-4, rtol=1e-4)
+
+                torch.testing.assert_close(torch_grad, thunder_grad, atol=1e-3, rtol=1e-3)
```
Comment (k223kim) [inline thunder/tests/distributed/helper.py]: @t-vi I think this should fail if thunder and torch are different

File: thunder/tests/distributed/helper.py (hunk 2 above)
Comment (crcrpar) [reply thunder/tests/distributed/helper.py]: from the perspective of what we test, there's no differences but making `use_no_sync` argument would make test logs way better. It clearly tells us which setting fails or not. I'd have `use_no_sync` as an argument as @Borda suggests.

Comment (t-vi) [reply thunder/tests/distributed/helper.py]: I think we should split the test.
That said, I would expect the no-sync test to roughly do the following:

```
for j in range(2):
    for i in range(2):
        with no_sync():
            fw
            loss
            bw
        grab grad.clones for comparison  # these would not be synced, even outside the context manager.
    fw
    loss
    bw  # sync happens only(!) here and below
    grab grad.clones for comparison  # these are synced
    zero_grad
    fw
    loss
    bw  # sync happens only(!) here and above
    grab grad.clones for comparison  # these are synced
    zero_grad
```

and then the comparison is between the gradients in order between plain PT DDP and Thunder DDP.

Comment (t-vi) [reply thunder/tests/distributed/helper.py]: Note that the grabbing of the grads happens outside the no_sync.
